### PR TITLE
FROM flywheel/hcp-base:0.2.0_4.0.1

### PR DESCRIPTION
**Pull request by joshicola**
This iteration of the hcp-diff gear utilizes a base image (flywheel/hcp-base:0.2.0_4.0.1) for the latest versions of all HCP (v4.0.1) requirements. It is an upgrade from the bash gear.

### CODE REVIEW

File: run.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    FSLSUBOPTIONS = "-l "+ LogFileDirFull
+
+    command_common=[
+        op.join(context.gear_dict['environ']['FSLDIR'],'bin','fsl_sub'),
+        FSLSUBOPTIONS
+    ]
+    
+    context.gear_dict['command_common'] = command_common
+
+    # Execute fMRI Volume Pipeline
```
Comment (margaretmahan) [inline run.py]: Replace comment with diffusion.

File: run.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    try:
+        DiffPreprocPipeline.execute(context)
+    except Exception as e:
+        context.log.exception(e)
+        context.log.fatal('The Diffusion Preprocessing Pipeline Failed!')
+        if context.config['save-on-error']:
+            results.cleanup(context)
+        os.sys.exit(1)
+
+    # Generate HCP-Functional QC Images
```
Comment (margaretmahan) [inline run.py]: Comment -> Diffusion

File: run.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    try:
+        hcpdiff_qc_mosaic.build(context)
+        hcpdiff_qc_mosaic.execute(context)
+    except Exception as e:
+        context.log.exception(e)
+        context.log.fatal('HCP Functional QC Images has failed!')
```
Comment (margaretmahan) [inline run.py]: Log message -> Diffusion

File: utils/args/DiffPreprocPipeline.py [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+def build(context):
+    config = context.config
+    inputs = context._invocation['inputs']
+
+    # Default Config Settings
+    # DwellTime for DWI volumes
+    EffectiveEchoSpacing = "NONE" 
+
+    # no gradient correction unless we are provided with a .grad file
```
Comment (margaretmahan) [inline utils/args/DiffPreprocPipeline.py]: Consider adding this to the if, making an if-else statement

File: utils/args/DiffPreprocPipeline.py [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+            os.makedirs(op.join(base_dir,'Pos{}'.format(i)), exist_ok=True)
+            os.makedirs(op.join(base_dir,'Neg{}'.format(i)), exist_ok=True)
+
+            # Grab the Phase Encoding
+            info_pos = inputs['DWIPositiveData{}'.format(j)]['object']['info']
+            info_neg = inputs['DWINegativeData{}'.format(j)]['object']['info']
+            if  ('PhaseEncodingDirection' in info_pos.keys()) and \
+                ('PhaseEncodingDirection' in info_neg.keys()):
+                test['PE']['Pos'] = tr("ijk", "xyz", info_pos['PhaseEncodingDirection'])
+                test['PE']['Neg'] = tr("ijk", "xyz", info_neg['PhaseEncodingDirection'])
```
Comment (margaretmahan) [inline utils/args/DiffPreprocPipeline.py]: Not certain on this, but should these be the same "ijk" and "xyz"?

File: utils/args/DiffPreprocPipeline.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    params['b0maxbval'] = b0maxbval
+    params['combine-data-flag'] = CombineDataFlag
+    params['extra-eddy-arg'] = ExtraEddyArgs
+    params['no-gpu'] = No_GPU
+        
+    params['printcom'] = ' '
+    context.gear_dict['Diff-params'] = params
+
+def validate(context):
+    # The validation code is repeated from "build()" to reinforce clarity of
```
Comment (dpark6060) [inline utils/args/DiffPreprocPipeline.py]: Idk....

File: Dockerfile [hunk 7]
```diff
@@ -2,188 +2,107 @@
...
+# HCP Pipelines v4.0.1 are installed in base image. Ensure variables are set
+
+# Set up specific environment variables for the HCP Pipeline
+ENV FSL_DIR="${FSLDIR}" \ 
+    HCPPIPEDIR=/opt/HCP-Pipelines \ 
+    MSMBINDIR=${HCPPIPEDIR}/MSMBinaries \ 
+    MSMCONFIGDIR=${HCPPIPEDIR}/MSMConfig
+#ENV MATLAB_COMPILER_RUNTIME=/media/myelin/brainmappers/HardDrives/1TB/MATLAB_Runtime/v901
+#ENV FSL_FIXDIR=/media/myelin/aahana/fix1.06
+
```
[the remark this answers is not in the document]
Comment (joshicola) [reply Dockerfile]: Nope... Crrrruuuuufffttt!!!!

File: utils/args/hcpdiff_qc_mosaic.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+# qc_scene_root="${StudyFolder}/${Subject}"
+
+# qc_outputdir="${StudyFolder}"
+# mkdir -p ${qc_outputdir}
+
+# qc_image_root="${qc_outputdir}/${Subject}_${DWIName}.hcpdiff_QC."
+
+# set -x
+# ${RUN_QC} ${SCRIPT_DIR}/hcpdiff_qc_mosaic.sh \
```
Comment (dpark6060) [inline utils/args/hcpdiff_qc_mosaic.py]: are these comments left in for reference?  Comment indicating so or remove

File: utils/args/hcpdiff_qc_mosaic.py [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+    params['qc_scene_root'] = op.join(context.work_dir, config['Subject'])
+    params['DWIName'] = config['DWIName']
+    params['qc_image_root'] = op.join(
+        context.work_dir,
+        config['Subject'] + '_' + config['DWIName'] + '.hcpdiff_QC.'
+    )
+    context.gear_dict['QC-Params'] = params
+
+def validate(context):
```
Comment (dpark6060) [inline utils/args/hcpdiff_qc_mosaic.py]: you sure you don't want to validate a 5th time? 😆

File: Dockerfile [hunk 10]
```diff
@@ -2,188 +2,107 @@
...
 
 #############################################
-# Download and install Connectome Workbench
-RUN apt-get update && apt-get -y install connectome-workbench=1.2.3-1~nd14.04+1
+# FreeSurfer is installed in base image. Ensure environment is set
+# 6.0.1 ftp://surfer.nmr.mgh.harvard.edu/pub/dist/freesurfer/6.0.1/freesurfer-Linux-centos6_x86_64-stable-pub-v6.0.1.tar.gz
+# 5.3.0 ftp://surfer.nmr.mgh.harvard.edu/pub/dist/freesurfer/5.3.0-HCP/freesurfer-Linux-centos4_x86_64-stable-pub-v5.3.0-HCP.tar.gz
 
-ENV CARET7DIR=/usr/bin
+# Set up the FreeSurfer environment
```
[the remark this answers is not in the document]
Comment (joshicola) [reply Dockerfile]: Nice catch....

These were copied from the original gear.... without much thought.

I like the consistency myself.

File: utils/diff_utils.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+import os, os.path as op
+import glob
+
+def configs_to_export(context):
+    """
+    Export HCP Diffusion Pipeline configuration into the Subject directory
+    Return the config and filename
+    """
+    config = {}
+    hcpdiff_config={'config': config}
```
Comment (dpark6060) [inline utils/diff_utils.py]: Yo dawg, I heard you love config, so I put your config in your config, so you can config while you config

File: Dockerfile [hunk 12]
```diff
@@ -2,188 +2,107 @@
...
+ENV OS Linux \ 
+    FS_OVERRIDE 0 \ 
+    FIX_VERTEX_AREA= \
```
[the remark this answers is not in the document]
Comment (joshicola) [reply Dockerfile]: FIX_VERTEX_AREA is only used for `recon-all`... and it only needs to be set.... to ANYTHING... in order for it to work.

https://surfer.nmr.mgh.harvard.edu/fswiki/PreviousReleaseNotes

File: manifest.json [hunk 13]
```diff
@@ -8,28 +8,32 @@
   "url": "https://github.com/Washington-University/Pipelines",
   "source": "https://github.com/flywheel-apps/hcp-diff",
   "cite": "Glasser M. F., Sotiropoulos S. N., Wilson J. A., Coalson T. S., Fischl B., Andersson J. L., … Consortium, W. U.-M. H. (2013). The minimal preprocessing pipelines for the Human Connectome Project. NeuroImage, 80, 105–124.",
-  "version": "0.1.5",
+  "version": "0.2.0_4.0.1",
```
[the remark this answers is not in the document]
Comment (joshicola) [reply manifest.json]: We had this conversation (on Slack).  Resulting in needing to escalate that to Michael.

Although, my understanding is:
```<gear version>_<HCP Pipelines version>```

File: run.py (hunk 1 above)
Comment (joshicola) [reply run.py]: Copied Code... Copied C0DE....

File: utils/args/DiffPreprocPipeline.py (hunk 4 above)
Comment (joshicola) [reply utils/args/DiffPreprocPipeline.py]: Right... Copied Code... without thought... so much of it!!!!

File: utils/args/DiffPreprocPipeline.py (hunk 5 above)
Comment (joshicola) [reply utils/args/DiffPreprocPipeline.py]: They should be the same, but opposite. ("x","x-").  This is tested in `validate`.

File: Dockerfile [hunk 14]
```diff
@@ -2,188 +2,107 @@
...
-    cd / && \
-    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
-
-ENV HCPPIPEDIR=/opt/HCP-Pipelines
-
-# Manual patch needed to add non-GPU option to DiffPreprocPipeline.sh
-# (See also pipeline_patch/patch_notes.txt)
-COPY pipeline_patch/ ${HCPPIPEDIR}/
+# MSM_HOCR v3 binary is installed in base image.
+ENV MSMBINDIR=${HCPPIPEDIR}/MSMBinaries
```
[the remark this answers is not in the document]
Comment (joshicola) [reply Dockerfile]: Yes... Good to be specific here.... 

Copied old code... thoughtless of me...

File: run.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python3
+import json
+import os, os.path as op
+from zipfile import ZipFile
+import traceback
+from utils import results, gear_preliminaries, diff_utils
```
[the remark this answers is not in the document]
Comment (joshicola) [reply run.py]: Yes....

File: run.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+from utils.args import DiffPreprocPipeline, hcpdiff_qc_mosaic
```
[the remark this answers is not in the document]
Comment (joshicola) [reply run.py]: No problem.  My vscode configuration Now does this AutoMagically!!!!

File: run.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+
+import flywheel
+
+def main():
+    # Preamble: take care of all gear-typical activities.
+    context = flywheel.GearContext()
+    context.gear_dict = {}
```
[the remark this answers is not in the document]
Comment (joshicola) [reply run.py]: Hey, It worked for what it was meant to do.

File: run.py [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+        os.sys.exit(1)
+
+    ############################################################################
+    ###################Build and Validate Parameters############################
+    # Doing as much parameter checking before ANY computation.
+    # Fail as fast as possible.
+
+    try:
+        # Build and validate from Volume Processing Pipeline
+        DiffPreprocPipeline.build(context)
```
[the remark this answers is not in the document]
Comment (joshicola) [reply run.py]: Yes.

Other than 'Foolish consistency is the Hobgoblin of ....':

It reflects the exact name and format of the command being executed.

File: run.py [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+        DiffPreprocPipeline.validate(context)
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (joshicola) [reply run.py]: Good question.

There is the validation of the config.json against the manifest.json is done in the UI but not in the SDK.  This ensures that the `single` parameters have validated values (ranges, etc...)

Whereas the `DiffPreprocPipeline.validate()` ensures an application-specific validation of these parameters in relationship to one another.

File: utils/args/DiffPreprocPipeline.py [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+    GradientDistortionCoeffs = "NONE"
+
+    if 'GradientCoeff' in inputs.keys():
+        GradientDistortionCoeffs = context.get_input_path('GradientCoeff')
+
+    # Set PEdir variable based on Phase-Encoding directions
+    PEdir = ""
+    pedir_pos = None
+    pedir_neg = None
+    if  ('DWIPositiveData' in inputs.keys()) and \
```
[the remark this answers is not in the document]
Comment (joshicola) [reply utils/args/DiffPreprocPipeline.py]: My comments could have been more clear.... and it was months ago.

The whole structure is re-established.  If there was a transferable datatype (via `context.gear_dict` I would have done it!!!! ;) )

File: utils/args/DiffPreprocPipeline.py (hunk 6 above)
Comment (joshicola) [reply utils/args/DiffPreprocPipeline.py]: I will make the rationale more clear.

File: utils/args/hcpdiff_qc_mosaic.py (hunk 8 above)
Comment (joshicola) [reply utils/args/hcpdiff_qc_mosaic.py]: Forgotten Cruft from another age.

File: utils/args/hcpdiff_qc_mosaic.py (hunk 9 above)
Comment (joshicola) [reply utils/args/hcpdiff_qc_mosaic.py]: Very funny :face_with_head_bandage: ...

I will put a comment here.

File: utils/diff_utils.py (hunk 11 above)
Comment (joshicola) [reply utils/diff_utils.py]: It is really weird... But this is the format it is looking for.

File: run.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (joshicola) [reply run.py]: See https://github.com/flywheel-apps/hcp-diff/blob/python-dev/utils/args/DiffPreprocPipeline.py#L346

**Pull request merged**